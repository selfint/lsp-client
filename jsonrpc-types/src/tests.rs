use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
pub(crate) struct Params {
    pub p0: u32,
    pub p1: u32,
}

#[cfg(test)]
macro_rules! snapshot {
    ($e:expr) => {
        insta::assert_json_snapshot!($e);

        let serialized = serde_json::to_value($e);
        let deserialized = serde_json::from_value(serialized.unwrap()).unwrap();

        similar_asserts::assert_eq!($e, deserialized);
    };
}

#[cfg(test)]
pub(crate) use snapshot;
