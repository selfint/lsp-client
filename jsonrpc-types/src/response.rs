use serde::{ser::SerializeStruct, Deserialize, Serialize};

use crate::JSONRPC_V2;

#[derive(Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct Response<R, E> {
    #[serde(flatten)]
    pub result: JsonRPCResult<R, E>,
    pub id: Option<u64>,
}

impl<R, E> Response<R, E> {
    pub fn new(result: JsonRPCResult<R, E>, id: Option<u64>) -> Self {
        Self { result, id }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum JsonRPCResult<R, E> {
    Result(R),
    Error(JsonRPCError<E>),
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct JsonRPCError<D> {
    pub code: i64,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<D>,
}

impl<R: Serialize, E: Serialize> Serialize for Response<R, E> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("Response", 3)?;
        state.serialize_field("jsonrpc", JSONRPC_V2)?;

        // flatten result
        match &self.result {
            JsonRPCResult::Result(r) => state.serialize_field("result", r)?,
            JsonRPCResult::Error(e) => state.serialize_field("error", e)?,
        };

        state.serialize_field("id", &self.id)?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::tests::{snapshot, Params};

    #[test]
    fn test_response_serde() {
        macro_rules! snapshot_permutations {
            ($data:expr) => {
                snapshot!(Response::new(
                    JsonRPCResult::<_, ()>::Result($data),
                    Some(1)
                ));
                snapshot!(Response::new(JsonRPCResult::<_, ()>::Result($data), None));
                snapshot!(Response::new(
                    JsonRPCResult::<(), _>::Error(JsonRPCError {
                        code: -1,
                        message: "message".to_string(),
                        data: Some($data)
                    }),
                    Some(1)
                ));
                snapshot!(Response::new(
                    JsonRPCResult::<(), ()>::Error(JsonRPCError {
                        code: -1,
                        message: "message".to_string(),
                        data: None
                    }),
                    Some(1)
                ));
                snapshot!(Response::new(
                    JsonRPCResult::<(), _>::Error(JsonRPCError {
                        code: -1,
                        message: "message".to_string(),
                        data: Some($data)
                    }),
                    None
                ));
                snapshot!(Response::new(
                    JsonRPCResult::<(), ()>::Error(JsonRPCError {
                        code: -1,
                        message: "message".to_string(),
                        data: None
                    }),
                    None
                ));
            };
        }

        snapshot_permutations!(1);
        snapshot_permutations!(vec![1, -1]);
        snapshot_permutations!(Params { p0: 0, p1: 1 });
    }
}
