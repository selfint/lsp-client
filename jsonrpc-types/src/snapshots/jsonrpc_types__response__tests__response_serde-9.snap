---
source: jsonrpc-types/src/response.rs
expression: "Response::new(JsonRPCResult::<(),\n            _>::Error(JsonRPCError {\n            code: -1,\n            message: \"message\".to_string(),\n            data: Some(vec![1, - 1]),\n        }), Some(1))"
---
{
  "jsonrpc": "2.0",
  "error": {
    "code": -1,
    "message": "message",
    "data": [
      1,
      -1
    ]
  },
  "id": 1
}
